In [1]:
import pandas as pd

In [3]:
results = pd.DataFrame({
    'pattern' : ['Center Squared', 'Center Squared', 'Center Squared', 
                 'Circle', 'Circle', 'Circle', 'Circle',
                 'Diamond', 'Diamond', 'Diamond', 'Diamond',
                 'Thick Right', 'Thick Right', 'Thick Right', 
                 'Thin Right', 'Thin Right', 'Thin Right', 'Thin Right'],
    'epoch' : [5000, 5000, 3000, 
               5000, 5000, 80000, 10000,
               5000, 5000, 5000, 2500,
               5000, 5000, 3000,
               5000, 5000, 3000, 2000
              ],
    'alpha' : [.001, .01, .005, 
               .001, .01, .001, .001,
               .001, .01, .005, .005,
               .001, .01, .01,
               .001, .01, .01, .01
              ],
    'loss' : [.1361, .1374, .1085, 
              .1525, .1314, .1249, .1139,
              .1340, .1175, .0756, .0951,
              .0832, .0443, .0488,
              .1124, .0306, .0364, .0430
             ],
    'acc' : [94.4, 94.8, 95.4, 
             95.2, 94.6, 96.6, 97.0, 
             97.6, 95.0, 99.6, 99.4, 
             97.4, 98.6, 98.4, 
             98.6, 99.2, 99.0, 99.0
            ]
    
})

results

pattern  epoch  alpha    loss   acc
0   Center Squared   5000  0.001  0.1361  94.4
1   Center Squared   5000  0.010  0.1374  94.8
2   Center Squared   3000  0.005  0.1085  95.4
3           Circle   5000  0.001  0.1525  95.2
4           Circle   5000  0.010  0.1314  94.6
5           Circle  80000  0.001  0.1249  96.6
6           Circle  10000  0.001  0.1139  97.0
7          Diamond   5000  0.001  0.1340  97.6
8          Diamond   5000  0.010  0.1175  95.0
9          Diamond   5000  0.005  0.0756  99.6
10         Diamond   2500  0.005  0.0951  99.4
11     Thick Right   5000  0.001  0.0832  97.4
12     Thick Right   5000  0.010  0.0443  98.6
13     Thick Right   3000  0.010  0.0488  98.4
14      Thin Right   5000  0.001  0.1124  98.6
15      Thin Right   5000  0.010  0.0306  99.2
16      Thin Right   3000  0.010  0.0364  99.0
17      Thin Right   2000  0.010  0.0430  99.0

Which patterns was the neural network able to learn quickly and which took longer? 

•	Based on the results from the table above the neural network learned the Thin Right pattern the quickest with Thick Right behind it.  

•	The neural network had the hardest times learning the Circle pattern, unlike the rest of the patterns that performed well with a learning rate of .01, when trying to use .01 for the Circle pattern their was a significant level of volatility in loss and accuracy values between iterations.  Furthermore, unlike the other patterns, we had to increase the number of epochs for the algorithm to fit the Circle pattern.


What learning rates and numbers of iterations worked well?

•	Centered Squared: A learning rate of .005 and 3000 iterations seemed to perform the best based on the lowest loss and highest accuracy.

•	Circle: We had to increase the number of iterations to 10,000, doubling the original epoch value, to see the loss and accuracy converge.  Any attempt to increase the learning rate resulted in high volatility in loss and accuracy scores.

•	Diamond: While the algorithm performed the best with 500 iterations and a learning rate of .005, I believe using 2500 iterations instead is better since the training time is cut in half with only a slight increase in loss and decrease of accuracy.

•	Thick Right: By dropping the number of iterations to 3000 and pushing the learning rate up to .01 converged to a loss value of .0488 and an accuracy of ~98%.

•	Thin Right: Like the Thick Right and Diamond pattern, we cut the training time down by reducing the epoch from the initial 5000 to 2000 – and having a learning rate of .01. 